cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaPool.java

```diff
@@ -0,0 +1,296 @@
+                        .withStrimziCluster(kafka.getMetadata().getName())
+                        .withStrimziComponentType(COMPONENT_TYPE)
+                        .withStrimziPoolName(pool.getMetadata().getName())
+                        // Kubernetes labels
+                        .withKubernetesName(COMPONENT_TYPE)
+                        .withKubernetesInstance(kafka.getMetadata().getName())
+                        .withKubernetesPartOf(kafka.getMetadata().getName())
+                        .withKubernetesManagedBy(STRIMZI_CLUSTER_OPERATOR_NAME),
+                ownerReference
+        );
+
+        this.poolName = pool.getMetadata().getName();
+        this.idAssignment = idAssignment;
+    }
+
+    /**
+     * Creates the Kafka pool model instance from a Kafka and KafkaNodePool CRs
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param kafka             Kafka custom resource
+     * @param pool              Node pool configuration
+     * @param idAssignment      Assignment of node IDs
+     * @param oldStorage        The current storage configuration (based on the info from Kubernetes cluster, not from
+     *                          the Kafka CR)
+     * @param ownerReference    Owner reference which should be used for this component. This should be the KafkaNodePool
+     *                          CR for regular pool or the Kafka CR for virtual node pool.
+     * @param useKRaft          Flag indicating if KRaft is enabled
+     *
+     * @return  Kafka pool instance
+     */
+    public static KafkaPool fromCrd(
+            Reconciliation reconciliation,
+            Kafka kafka,
+            KafkaNodePool pool,
+            NodeIdAssignment idAssignment,
+            Storage oldStorage,
+            OwnerReference ownerReference,
+            boolean useKRaft
+    ) {
+        ModelUtils.validateComputeResources(pool.getSpec().getResources(), "KafkaNodePool.spec.resources");
+        StorageUtils.validatePersistentStorage(pool.getSpec().getStorage(), "KafkaNodePool.spec.storage");
+
+        KafkaPool result = new KafkaPool(reconciliation, kafka, pool, componentName(kafka, pool), ownerReference, idAssignment);
+
+        result.gcLoggingEnabled = isGcLoggingEnabled(kafka, pool);
+        result.jvmOptions = pool.getSpec().getJvmOptions() != null ? pool.getSpec().getJvmOptions() : kafka.getSpec().getKafka().getJvmOptions();
+        result.resources = pool.getSpec().getResources() != null ? pool.getSpec().getResources() : kafka.getSpec().getKafka().getResources();
+
+        if (useKRaft)   {
+            // KRaft is enabled. The roles are what is set in the KafkaNodePool CR or mixed node by default
+            result.kraftRoles = pool.getSpec().getRoles() != null && !pool.getSpec().getRoles().isEmpty() ? new HashSet<>(pool.getSpec().getRoles()) : Set.of(ProcessRoles.CONTROLLER, ProcessRoles.BROKER);
```

Comment:
I think it would make sense to have them required for KRaft. But with ZooKeeper, not sure it makes sense to force this on users to set the role to broker. In the CRD schema, we cannot require the roles only in KRaft. So if we set them as required, it would be required even outside of KRaft -> but without CRD based validation that without a KRaft you have always only `BROKER` selected. 

Is that what we really want? If you think it should be required despite this, I'm fine to change it.

Reply:
The cost of making the roles required in the CRD today, AFAIU, is an instruction to the user in the docs for upgrading their ZK cluster to NodePools (and also for creating ZK clusters on NP). That seems small to me. The benefit is we end up with CRs which are much more explicit. So we're trading off ease of use of NodePools with ZK now against requiring explictness when using NodePools in a KRaft only future. One thing about the future is that there's a lot of it, so personally I'm inclined to say let's make it required today.